DOCS-2691: Add Create a Sensor Module how-to (#3259)
viamrobotics/docs: docs/how-tos/sensor-module.md

```python {class="line-numbers linkable-line-numbers"}
# my-sensor-test.py
from serial import Serial
import time

def main():

    port = Serial('/dev/ttyAMA0', baudrate=9600)

    def parse_data(data):

        if len(data) < 2:
            return {}
        if data[0] == 0x42 and data[1] == 0x4d:
            pm2_5_atm = (data[12] << 8) + data[13]
            pm10_atm = (data[14] << 8) + data[15]
            print(f'PM2.5 (atmospheric): {pm2_5_atm} µg/m3')
            print(f'PM10 (atmospheric): {pm10_atm} µg/m3')

            # Return a dictionary of the readings
            return {
                "pm2_5_atm": pm2_5_atm,
                "pm10_atm": pm10_atm,
            }

        else:
            print('Data does not start with the expected start bytes.')
            return {}

    while port.in_waiting == 0:
        time.sleep(0.01)  # wait for 10 ms

    data = port.read(port.in_waiting)
    print(parse_data(data))

if __name__ == "__main__":
    main()
```

{{< /expand >}}

Run your test script from your terminal and make sure you are able to get readings from the sensor before proceeding.

## Generate boilerplate module code

> You need some boilerplate files to make your module.
> To make this easier, use the Viam module generator:

Author: ```suggestion
There are a few standardized files that must be part of any module.
You can create these automatically using the Viam module generator:
```

Making a module to support your sensor will allow you to use it with Viam's data capture and sync tools, as well as using the sensor API (using any of the different programming language [SDKs](/sdks/)) to get readings from it.

{{% alert title="In this page" color="info" %}}

1. [Start with a test script](#start-with-a-test-script)
1. [Generate boilerplate module code](#generate-boilerplate-module-code)
1. [Implement the sensor API](#implement-the-sensor-api)
1. [Make the module executable](#make-the-module-executable)
1. [Test your module locally](#test-your-module-locally)
1. [Upload your module](#upload-your-module-to-the-registry)

{{% /alert %}}

## Start with a test script

Start by getting a test script working so you can check that the sensor code itself works before packaging it into a module.

Since this how-to uses Python, you need a Python test script so that you can more easily wrap it in a Python-based module.

> You'll still be able to use any of Viam's SDKs with the module.

Reviewer: ```suggestion
You'll still be able to use any of Viam's SDKs to query data from machines that use the module.
```

Author: Would you call it querying if it's getting readings directly from the machine as opposed to from the database?

Reviewer: ```suggestion
You'll still be able to use any of Viam's SDKs to get readings from machines that use the module.
```

> This step depends completely on your sensor hardware (or software)--just find or write some script that gets readings from the sensor and prints them out.

Reviewer: ```suggestion
This step depends completely on your sensor hardware (or software) - just find or write some script that gets readings from the sensor and prints them out.
```

Author: Why this change? An m dash is grammatically correct here no?

Reviewer: I don't have strong opinions on emdashes though I personally tend to use the regular dash... Let's usee the proper emdash then (rather than two dashes)

```suggestion
This step depends completely on your sensor hardware (or software)—just find or write some script that gets readings from the sensor and prints them out.
```

Author: Two dashes renders as an m dash. One dash renders as an n dash. This is how two dashes renders; it is correct:
<img width="111" alt="image" src="https://github.com/user-attachments/assets/d2c24d32-a6e7-45e5-bb03-8d21db01698d">

Reviewer: also with `—` (my second suggestion) because that's different from `-`

Reviewer: I just tested this and the emdash works so please apply the second suggestion

Author: Sorry I should have three dashes not two to make an em dash. I was mixing en dashes and hyphens.
```suggestion
This step depends completely on your sensor hardware (or software)---just find or write some script that gets readings from the sensor and prints them out.
```

# The order of variables in hourly or daily is important
# to assign them correctly below
url = "https://air-quality-api.open-meteo.com/v1/air-quality"
params = {
  "latitude": 44.0582,
  "longitude": -121.3153,
  "current": ["pm10", "pm2_5"],
  "timezone": "America/Los_Angeles"
}
responses = openmeteo.weather_api(url, params=params)

# Process first location.
response = responses[0]
print(f"Coordinates {response.Latitude()}°N {response.Longitude()}°E")
print(f"Elevation {response.Elevation()} m asl")
print(f"Timezone {response.Timezone()} {response.TimezoneAbbreviation()}")
print(f"Timezone difference to GMT+0 {response.UtcOffsetSeconds()} s")

# Current values. The order of variables needs to be the same as requested.
current = response.Current()
current_pm10 = current.Variables(0).Value()
current_pm2_5 = current.Variables(1).Value()

print(f"Current time {current.Time()}")
print(f"Current pm10 {current_pm10}")
print(f"Current pm2_5 {current_pm2_5}")
```

{{< /expand >}}

{{< expand "An example of getting PM2.5 and PM10 readings over serial from sensor hardware" >}}
This script is based on the [pms7003 module](https://app.viam.com/module/joyce/pms7003).

> Before creating the module, the following script could have been used to verify the code that gets the readings from the sensor.

Reviewer: It can still be used so let's instead point out that this is something they can do to explore
```suggestion
For most modules, the implementation code is linked from the module page in the registry, so you can see how it works.
```

Author: This is not the code that is in the repo. This is a modified version of it, that prints things instead of logging them, and puts the main functionality directly into main() so that the script can be run independently.

Author: However, I have not actually tested this script since I do not have this hardware. I could ax this example and include only the first one (which I wrote and tested).

Reviewer: looks like it should run to me, I think that's an ok risk to take.
I think it's still true to say the script is based on that module? You did copy the code from there. And it is useful to look at how people write their modules.

Author: Yeah, the sentence before this one still says that. I agree that is useful, but probably more in the implementation step. I'll tweak line 335 to be more explicit about that.

Reviewer: That works. I've added a suggestion there. I'd still like this sentence gone.
```suggestion
```

Reviewer: Why did you resolve this? Sentence is still there.

Author: Didn't mean to--may have gotten auto-resolved during rebase??
I'd also been confused because I added the sentence after line 335 to this effect and hadn't realized you'd wanted this deleted and your specific wording used. Committing now.

```python {class="line-numbers linkable-line-numbers"}
# my-sensor-test.py
from serial import Serial
import time

def main():

    port = Serial('/dev/ttyAMA0', baudrate=9600)

    def parse_data(data):

        if len(data) < 2:
            return {}
        if data[0] == 0x42 and data[1] == 0x4d:
            pm2_5_atm = (data[12] << 8) + data[13]
            pm10_atm = (data[14] << 8) + data[15]
            print(f'PM2.5 (atmospheric): {pm2_5_atm} µg/m3')
            print(f'PM10 (atmospheric): {pm10_atm} µg/m3')

            # Return a dictionary of the readings
            return {
                "pm2_5_atm": pm2_5_atm,
                "pm10_atm": pm10_atm,
            }

        else:
            print('Data does not start with the expected start bytes.')
            return {}

    while port.in_waiting == 0:
        time.sleep(0.01)  # wait for 10 ms

    data = port.read(port.in_waiting)
    print(parse_data(data))

if __name__ == "__main__":
    main()
```

{{< /expand >}}

> {{< expand "Another example (this one is for a base, not a sensor)" >}}
> 
> An example from the [custom robot dog base tutorial](/tutorials/custom/custom-base-dog/).
> When you run this with the robot dog hardware, you can tell it works because the robot dog takes a few steps forward.
> 
> ```python {class="line-numbers linkable-line-numbers"}
> # dog_test.py is for testing the connection
> import socket
> import time
> 
> s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
> s.connect(("PASTE DOG IP ADDRESS HERE", 5001))
> 
> cmd = "CMD_MOVE_FORWARD#15"
> s.send(cmd.encode("utf-8"))
> time.sleep(7)
> cmd = "CMD_MOVE_STOP"
> s.send(cmd.encode("utf-8"))
> cmd = "CMD_RELAX"
> s.send(cmd.encode("utf-8"))
> ```
> 
> {{< /expand >}}

Reviewer: I think that's more confusing than helpful since it's not a sensor I thought you maybe were getting data from the base and using that as a sensor but I think not
```suggestion
```

Author: Yeah I was wondering about whether to include this. I left it in in case people come to this page to create a Python module for not-a-sensor, but okay will remove!

Run your test script from your terminal and make sure you are able to get readings from the sensor before proceeding.

## Generate boilerplate module code

You need some boilerplate files to make your module.
To make this easier, use the Viam module generator:

1. Follow the steps in the [readme](https://github.com/viam-labs/generator-viam-module/tree/main) to install the generator and run it.

1. When prompted for a model triplet, use `<your organization public namespace>:<repo name>:<what you want to call your sensor model>`.
   For example, `jessamy:weather:meteo_PM`.

   - You can find your organization namespace by going to your organization settings in the [Viam app](https://app.viam.com).

>    - The repo name (family name) is the name of the GitHub repo where you will put your module code.

Reviewer: ```suggestion
   - The repo name (also used as family name) is generally the name of the GitHub repo where you will put your module code.
```

Author: We say repo name elsewhere in the docs, but the generator says family name. What exactly does "also used as" mean here?

   For more information, see [Name your new resource model](/how-tos/create-module/#name-your-new-resource-model).

1. For the API triplet, enter `rdk:component:sensor`.
   This means that you are implementing the standard Viam sensor API.

1. When asked whether this is a Viam SDK built-in API, enter `yes`.

The generator creates a `run.sh` file, a `requirements.txt` file, a readme, and source code files.
In the next section, you'll customize some of these files to support your sensor.

## Implement the sensor API

Other than the inherited methods, the [sensor API](/components/sensor/#api) only contains one method, `GetReadings()`.
You need to implement this method to satisfy the sensor API:

1. In the generated <file>/YOUR_MODULE_NAME/src/</file> directory, open the </file>MODEL_NAME.py</file> file.

1. Edit the config attributes to fit your sensor.
   For example, if your sensor requires two pins, copy the `some_pin` lines and add another pin with a different name.
   If you want to be able to configure something else, for example the location to get online data from, you can add attributes for that.

>    If your sensor doesn't require any configuration, delete the `some_pin` lines but don't delete the `validate` and `reconfigure` functions entirely; they're needed for the module to function even if they don't do anything interesting.

Reviewer: ```suggestion
   If your sensor doesn't require any configuration, delete the `some_pin` lines but don't delete the `validate` and `reconfigure` functions entirely; they're needed for the module to function even if they don't actually validate the input or reconfigure the resource.
```

        current_pm2_5 = current.Variables(1).Value()

        LOGGER.info(current_pm2_5)

        # Return a dictionary of the readings
        return {
            "pm2_5": current_pm2_5,
            "pm10": current_pm10
        }
```

{{< /expand >}}

For more examples, see the [existing sensor models](/components/sensor/#available-models).

## Edit requirements.txt

Update the generated <file>requirements.txt</file> file to include any packages that must be installed for the module to run.
Depending on your use case, you may not need to add anything here beyond `viam-sdk` which is auto-populated.

## Make the module executable

You need an executable file so that `viam-server` can run your module.
The module generator already created the <file>run.sh</file> "entrypoint" file for you, so all you need to do is make this file executable by running the following command with your correct file path:

```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
sudo chmod +x <your-file-path-to>/run.sh
```

## Test your module locally

It's a good idea to test your module locally before uploading it to the [Viam registry](https://app.viam.com/registry):

> 1. Follow [the local module testing instructions](/how-tos/create-module/#test-your-module-locally) to configure your local module and then the associated model on your machine.

Reviewer: I think those steps are so small state you can include them directly, maybe in slightly shortened form

   ![Configuring a local model after the local module is configured.](/how-tos/sensor-module-config.png)

1. Click the **TEST** bar at the bottom of your sensor configuration, and check whether readings are being returned there.

   ![The control tab with pm10 and pm2_5 readings displayed.](/how-tos/sensor-test.png)

If it works, you're almost ready to share your module by uploading it to the registry.

> If not, you have some debugging to do.

Reviewer: Maybe also direct them at the logs tab.
Which reminds me - you forgot prerequisites (a machine)

Author: Someone can create a module without having a machine. Granted they wouldn't be able to test the module but they could totally create and upload it. You think I should add it?

Author: I added it as a prereq to this section since I think it makes more sense that way

## Configure your `meteo_PM` sensor

Navigate to the **CONFIGURE** tab of your robot’s page in [the Viam app](https://app.viam.com/).
Add a component.
Select the `sensor` type, then select the `jessamy:weather:meteo_PM` model.
Enter a name for your sensor and click **Create**.

On the new component panel, copy and paste the following attribute template into your sensor’s **Attributes** box:

```json
{
  "latitude": <float>,
  "longitude": <float>
}
```

### Attributes

The following attributes are available for `rdk:sensor:jessamy:weather:meteo_PM` sensors:

| Name        | Type  | Inclusion | Description                            |
| ----------- | ----- | --------- | -------------------------------------- |
| `latitude`  | float | Optional  | Latitude at which to get the readings  |
| `longitude` | float | Optional  | Longitude at which to get the readings |

### Example Configuration

```json
{
  "latitude": -40.6,
  "longitude": 93.125
}
```
````

{{< /expand >}}

## Upload your module to the registry

> To share your module with others in your organization or with the world, [follow these instructions to upload your module to the modular resource registry](/how-tos/create-module/#upload-your-module-to-the-modular-resource-registry).

Reviewer: This should probably be its own how-to that we can then link to from any how to create a module how-to

Author: mm ok. I'll make a ticket

Before creating the module, the following script could have been used to verify the code that gets the readings from the sensor.

```python {class="line-numbers linkable-line-numbers"}
# my-sensor-test.py
from serial import Serial
import time

def main():

    port = Serial('/dev/ttyAMA0', baudrate=9600)

    def parse_data(data):

        if len(data) < 2:
            return {}
        if data[0] == 0x42 and data[1] == 0x4d:
            pm2_5_atm = (data[12] << 8) + data[13]
            pm10_atm = (data[14] << 8) + data[15]
            print(f'PM2.5 (atmospheric): {pm2_5_atm} µg/m3')
            print(f'PM10 (atmospheric): {pm10_atm} µg/m3')

            # Return a dictionary of the readings
            return {
                "pm2_5_atm": pm2_5_atm,
                "pm10_atm": pm10_atm,
            }

        else:
            print('Data does not start with the expected start bytes.')
            return {}

    while port.in_waiting == 0:
        time.sleep(0.01)  # wait for 10 ms

    data = port.read(port.in_waiting)
    print(parse_data(data))

if __name__ == "__main__":
    main()
```

{{< /expand >}}
{{% /tablestep %}}
{{% tablestep %}}

> **2. Run your test script**

Reviewer: You can if you want to skip having steps for this part. I agree that here it feels a bit unnecessary. In the rest it's more helpful. Up to you.

Author: Sorry i'm not sure what this comment is in relation to

Reviewer: I'm saying you don't need to put tablesteps here into the "Start with a test script" section if you don't want to

Author: oooh

Run your test script from your terminal and make sure you are able to get readings from the sensor before proceeding.

{{% /tablestep %}}
{{< /table >}}

## Generate boilerplate module code

There are a few standardized files that must be part of any module.
You can create these automatically using the Viam module generator:

{{< table >}}
{{% tablestep %}}

> **1. Install and run the module generator**

Reviewer: ```suggestion
**1. Install the module generator**
```

> Follow the steps in the [readme](https://github.com/viam-labs/generator-viam-module/tree/main) to install the generator and run it.

Reviewer: ```suggestion
The following commands will install a module generator [generator-viam-module](https://github.com/viam-labs/generator-viam-module/tree/main).
You must have node 16 or newer installed.

```sh
npm install -g yo generator-viam-module
```
```

{{% /tablestep %}}
{{% tablestep %}}
**2. Name your model**

When prompted for a model triplet, use `<your organization public namespace>:<repo name>:<what you want to call your sensor model>`.
For example, `jessamy:weather:meteo_PM`.

- You can find your organization namespace by going to your organization settings in the [Viam app](https://app.viam.com).

> - The repo name (family name) is the name of the GitHub repo where you will put your module code.

Reviewer: ```suggestion
- The repo name (also called family name) is generally the name of the GitHub repo where you will put your module code.
```

        cache_session = requests_cache.CachedSession(
          '.cache', expire_after=3600)
        retry_session = retry(cache_session, retries=5, backoff_factor=0.2)
        openmeteo = openmeteo_requests.Client(session=retry_session)

        # The order of variables in hourly or daily is
        # important to assign them correctly below
        url = "https://air-quality-api.open-meteo.com/v1/air-quality"
        params = {
            "latitude": self.latitude,
            "longitude": self.longitude,
            "current": ["pm10", "pm2_5"],
            "timezone": "America/Los_Angeles"
        }
        responses = openmeteo.weather_api(url, params=params)

        # Process location
        response = responses[0]

        # Current values. The order of variables needs
        # to be the same as requested.
        current = response.Current()
        current_pm10 = current.Variables(0).Value()
        current_pm2_5 = current.Variables(1).Value()

        LOGGER.info(current_pm2_5)

        # Return a dictionary of the readings
        return {
            "pm2_5": current_pm2_5,
            "pm10": current_pm10
        }
```

{{< /expand >}}

> For more examples, see the [existing sensor models](/components/sensor/#available-models).

Reviewer: ```suggestion
For more examples, go to the [Viam Registry](https://app.viam.com/registry). 
Most modules have their implementation code linked on their module page, so you can see how they work.
```

Author: You'd like me to do this and delete the sentence after it?

Reviewer: Yes

Author: ```suggestion
For more examples, see other sensor modules in the [Viam Registry](https://app.viam.com/registry). 
Most modules have their implementation code linked on their module page, so you can see how they work.
```

        cache_session = requests_cache.CachedSession(
          '.cache', expire_after=3600)
        retry_session = retry(cache_session, retries=5, backoff_factor=0.2)
        openmeteo = openmeteo_requests.Client(session=retry_session)

        # The order of variables in hourly or daily is
        # important to assign them correctly below
        url = "https://air-quality-api.open-meteo.com/v1/air-quality"
        params = {
            "latitude": self.latitude,
            "longitude": self.longitude,
            "current": ["pm10", "pm2_5"],
            "timezone": "America/Los_Angeles"
        }
        responses = openmeteo.weather_api(url, params=params)

        # Process location
        response = responses[0]

        # Current values. The order of variables needs
        # to be the same as requested.
        current = response.Current()
        current_pm10 = current.Variables(0).Value()
        current_pm2_5 = current.Variables(1).Value()

        LOGGER.info(current_pm2_5)

        # Return a dictionary of the readings
        return {
            "pm2_5": current_pm2_5,
            "pm10": current_pm10
        }
```

{{< /expand >}}

> For more examples, see other sensor modules in the [Viam Registry](https://app.viam.com/registry). 

Author: ```suggestion
For more examples, see other sensor modules in the [Viam Registry](https://app.viam.com/registry).
```

        retry_session = retry(cache_session, retries=5, backoff_factor=0.2)
        openmeteo = openmeteo_requests.Client(session=retry_session)

        # The order of variables in hourly or daily is
        # important to assign them correctly below
        url = "https://air-quality-api.open-meteo.com/v1/air-quality"
        params = {
            "latitude": self.latitude,
            "longitude": self.longitude,
            "current": ["pm10", "pm2_5"],
            "timezone": "America/Los_Angeles"
        }
        responses = openmeteo.weather_api(url, params=params)

        # Process location
        response = responses[0]

        # Current values. The order of variables needs
        # to be the same as requested.
        current = response.Current()
        current_pm10 = current.Variables(0).Value()
        current_pm2_5 = current.Variables(1).Value()

        LOGGER.info(current_pm2_5)

        # Return a dictionary of the readings
        return {
            "pm2_5": current_pm2_5,
            "pm10": current_pm10
        }
```

{{< /expand >}}

For more examples, see the [existing sensor models](/components/sensor/#available-models).

> You can find the code in the repos linked from each module's page in the registry.

Author: ```suggestion
```
